// @packages/app/dist/assets/cypress-dark-BnEPL-Nl.js
import { o as openBlock, i as createElementBlock, a as createBaseVNode, n as normalizeClass, d as defineComponent, g as useI18n, J as gql, l as computed, aj as useMutation, bF as VerticalBrowserListItems_SetBrowserDocument, B as renderList, b as unref, t as toDisplayString, V as createTextVNode, j as createCommentVNode, f as createVNode, w as withCtx, F as Fragment, aP as __unplugin_components_0$1, ae as useScreenshotStore, m as renderSlot } from "./index-aHhBZXEW.js";
import { _ as _sfc_main$2 } from "./Tooltip.vue_vue_type_style_index_0_lang-CSb17MEq.js";
const _hoisted_1$2 = {
  style: { "min-width": "16px", "min-height": "16px" },
  width: "16",
  height: "16",
  viewBox: "0 0 16 16",
  fill: "none",
  xmlns: "http://www.w3.org/2000/svg"
};
function render$1(_ctx, _cache) {
  return openBlock(), createElementBlock("svg", _hoisted_1$2, [
    createBaseVNode("path", {
      d: "M13 2H4C3.44772 2 3 2.44772 3 3V11H4H13V2Z",
      class: normalizeClass(["background", "icon-light"]),
      fill: "#D0D2E0",
      style: {}
    }),
    createBaseVNode("path", {
      d: "M13 11V2H4M13 11H4M13 11V14H4M3 11V13C3 13.5523 3.44772 14 4 14V14M3 11V3C3 2.44772 3.44772 2 4 2V2M3 11H4M4 2V11M4 11V14M10 5H7",
      class: normalizeClass(["border", "icon-dark"]),
      stroke: "#1B1E2E",
      "stroke-width": "2",
      "stroke-linecap": "round",
      "stroke-linejoin": "round"
    })
  ]);
}
const BookIcon = { name: "cy-book_x16", render: render$1 };
const _hoisted_1$1 = {
  style: { "min-width": "24px", "min-height": "24px" },
  width: "24",
  height: "24",
  viewBox: "0 0 24 24",
  fill: "none",
  xmlns: "http://www.w3.org/2000/svg"
};
const _hoisted_2$1 = /* @__PURE__ */ createBaseVNode("path", {
  d: "M20.4853 20.4853C15.799 25.1716 8.20101 25.1716 3.51472 20.4853C-1.17157 15.799 -1.17157 8.20101 3.51472 3.51472C8.20101 -1.17157 15.799 -1.17157 20.4853 3.51472C25.1716 8.20101 25.1716 15.799 20.4853 20.4853Z",
  fill: "#FFF",
  class: "icon-dark"
}, null, -1);
const _hoisted_3$1 = /* @__PURE__ */ createBaseVNode("path", {
  d: "M15.999 9L11 15L8.00098 12",
  stroke: "white",
  "stroke-width": "2",
  "stroke-linecap": "round",
  "stroke-linejoin": "round",
  class: "icon-light"
}, null, -1);
const _hoisted_4$1 = /* @__PURE__ */ createBaseVNode("circle", {
  cx: "12",
  cy: "12",
  r: "12",
  stroke: "black",
  "stroke-width": "1",
  style: { "stroke": "rgba(0,0,0,0.1)" }
}, null, -1);
const _hoisted_5$1 = [
  _hoisted_2$1,
  _hoisted_3$1,
  _hoisted_4$1
];
function render(_ctx, _cache) {
  return openBlock(), createElementBlock("svg", _hoisted_1$1, _hoisted_5$1);
}
const __unplugin_components_0 = { name: "cy-circle-check_x24", render };
const chromeIcon = "data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20viewBox='-10%20-10%20276%20276'%3e%3clinearGradient%20id='a'%20x1='145'%20x2='34'%20y1='253'%20y2='61'%20gradientUnits='userSpaceOnUse'%3e%3cstop%20offset='0'%20stop-color='%231e8e3e'/%3e%3cstop%20offset='1'%20stop-color='%2334a853'/%3e%3c/linearGradient%3e%3clinearGradient%20id='b'%20x1='111'%20x2='222'%20y1='254'%20y2='62'%20gradientUnits='userSpaceOnUse'%3e%3cstop%20offset='0'%20stop-color='%23fcc934'/%3e%3cstop%20offset='1'%20stop-color='%23fbbc04'/%3e%3c/linearGradient%3e%3clinearGradient%20id='c'%20x1='17'%20x2='239'%20y1='80'%20y2='80'%20gradientUnits='userSpaceOnUse'%3e%3cstop%20offset='0'%20stop-color='%23d93025'/%3e%3cstop%20offset='1'%20stop-color='%23ea4335'/%3e%3c/linearGradient%3e%3ccircle%20cx='128'%20cy='128'%20r='64'%20fill='%23fff'/%3e%3cpath%20fill='url(%23a)'%20d='M96%20183.4A63.7%2063.7%200%200%201%2072.6%20160L17.2%2064A128%20128%200%200%200%20128%20256l55.4-96A64%2064%200%200%201%2096%20183.4Z'/%3e%3cpath%20fill='url(%23b)'%20d='M192%20128a63.7%2063.7%200%200%201-8.6%2032L128%20256A128%20128%200%200%200%20238.9%2064h-111a64%2064%200%200%201%2064%2064Z'/%3e%3ccircle%20cx='128'%20cy='128'%20r='52'%20fill='%231a73e8'/%3e%3cpath%20fill='url(%23c)'%20d='M96%2072.6a63.7%2063.7%200%200%201%2032-8.6h110.8a128%20128%200%200%200-221.7%200l55.5%2096A64%2064%200%200%201%2096%2072.6Z'/%3e%3c/svg%3e";
const firefoxIcon = "" + new URL("firefox-1bWoP6pv.svg", import.meta.url).href;
const edgeIcon = "data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20xmlns:x='http://www.w3.org/1999/xlink'%20viewBox='0%200%2027600%2027600'%3e%3clinearGradient%20id='a'%20gradientUnits='userSpaceOnUse'/%3e%3clinearGradient%20id='b'%20x1='6870'%20x2='24704'%20y1='18705'%20y2='18705'%20x:href='%23a'%3e%3cstop%20offset='0'%20stop-color='%230c59a4'/%3e%3cstop%20offset='1'%20stop-color='%23114a8b'/%3e%3c/linearGradient%3e%3clinearGradient%20id='c'%20x1='16272'%20x2='5133'%20y1='10968'%20y2='23102'%20x:href='%23a'%3e%3cstop%20offset='0'%20stop-color='%231b9de2'/%3e%3cstop%20offset='.16'%20stop-color='%231595df'/%3e%3cstop%20offset='.67'%20stop-color='%230680d7'/%3e%3cstop%20offset='1'%20stop-color='%230078d4'/%3e%3c/linearGradient%3e%3cradialGradient%20id='d'%20cx='16720'%20cy='18747'%20r='9538'%20x:href='%23a'%3e%3cstop%20offset='.72'%20stop-opacity='0'/%3e%3cstop%20offset='.95'%20stop-opacity='.53'/%3e%3cstop%20offset='1'/%3e%3c/radialGradient%3e%3cradialGradient%20id='e'%20cx='7130'%20cy='19866'%20r='14324'%20gradientTransform='matrix(.14843%20-.98892%20.79688%20.1196%20-8759%2025542)'%20x:href='%23a'%3e%3cstop%20offset='.76'%20stop-opacity='0'/%3e%3cstop%20offset='.95'%20stop-opacity='.5'/%3e%3cstop%20offset='1'/%3e%3c/radialGradient%3e%3cradialGradient%20id='f'%20cx='2523'%20cy='4680'%20r='20243'%20gradientTransform='matrix(-.03715%20.99931%20-2.12836%20-.07913%2013579%203530)'%20x:href='%23a'%3e%3cstop%20offset='0'%20stop-color='%2335c1f1'/%3e%3cstop%20offset='.11'%20stop-color='%2334c1ed'/%3e%3cstop%20offset='.23'%20stop-color='%232fc2df'/%3e%3cstop%20offset='.31'%20stop-color='%232bc3d2'/%3e%3cstop%20offset='.67'%20stop-color='%2336c752'/%3e%3c/radialGradient%3e%3cradialGradient%20id='g'%20cx='24247'%20cy='7758'%20r='9734'%20gradientTransform='matrix(.28109%20.95968%20-.78353%20.22949%2024510%20-16292)'%20x:href='%23a'%3e%3cstop%20offset='0'%20stop-color='%2366eb6e'/%3e%3cstop%20offset='1'%20stop-color='%2366eb6e'%20stop-opacity='0'/%3e%3c/radialGradient%3e%3cpath%20id='h'%20d='M24105%2020053a9345%209345%200%2001-1053%20472%2010202%2010202%200%2001-3590%20646c-4732%200-8855-3255-8855-7432%200-1175%20680-2193%201643-2729-4280%20180-5380%204640-5380%207253%200%207387%206810%208137%208276%208137%20791%200%201984-230%202704-456l130-44a12834%2012834%200%20006660-5282c220-350-168-757-535-565z'/%3e%3cpath%20id='i'%20d='M11571%2025141a7913%207913%200%2001-2273-2137%208145%208145%200%2001-1514-4740%208093%208093%200%20013093-6395%208082%208082%200%20011373-859c312-148%20846-414%201554-404a3236%203236%200%20012569%201297%203184%203184%200%2001636%201866c0-21%202446-7960-8005-7960-4390%200-8004%204166-8004%207820%200%202319%20538%204170%201212%205604a12833%2012833%200%20007684%206757%2012795%2012795%200%20003908%20610c1414%200%202774-233%204045-656a7575%207575%200%2001-6278-803z'/%3e%3cpath%20id='j'%20d='M16231%2015886c-80%20105-330%20250-330%20566%200%20260%20170%20512%20472%20723%201438%201003%204149%20868%204156%20868a5954%205954%200%20003027-839%206147%206147%200%20001133-850%206180%206180%200%20001910-4437c26-2242-796-3732-1133-4392-2120-4141-6694-6525-11668-6525-7011%200-12703%205635-12798%2012620%2047-3654%203679-6605%207996-6605%20350%200%202346%2034%204200%201007%201634%20858%202490%201894%203086%202921%20618%201067%20728%202415%20728%202952s-271%201333-780%201990z'/%3e%3cuse%20fill='url(%23b)'%20x:href='%23h'/%3e%3cuse%20fill='url(%23d)'%20opacity='.35'%20x:href='%23h'/%3e%3cuse%20fill='url(%23c)'%20x:href='%23i'/%3e%3cuse%20fill='url(%23e)'%20opacity='.4'%20x:href='%23i'/%3e%3cuse%20fill='url(%23f)'%20x:href='%23j'/%3e%3cuse%20fill='url(%23g)'%20x:href='%23j'/%3e%3c/svg%3e";
const electronIcon = "data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20fill='%2347848f'%20viewBox='0%200%201610%201610'%3e%3cpath%20d='M603%20382c-195-36-349%202-411%20108-46%2079-33%20185%2032%20297%205%2010%2018%2013%2027%207%2010-5%2013-18%207-27-58-101-69-192-32-257%2051-88%20189-122%20369-89%2011%202%2021-5%2023-16s-4-21-15-23zM326%20924a1230%201230%200%2000298%20235c283%20163%20584%20207%20733%20106%209-6%2011-19%205-28s-19-11-28-5c-133%2091-419%2049-690-107a1200%201200%200%2001-289-228%2020%2020%200%2000-28-1%2020%2020%200%2000-1%2028z'/%3e%3cpath%20d='M1309%20900c128-151%20172-303%20111-409-45-78-141-119-267-121-11%200-20%209-20%2020s9%2020%2020%2020c114%201%20196%2037%20233%20101%2051%2088%2011%20224-107%20363a20%2020%200%20002%2028%2020%2020%200%200028-2zM981%20388a1263%201263%200%2000-356%20141c-293%20169-483%20418-457%20598a21%2021%200%200023%2017%2021%2021%200%200017-23c-24-160%20157-395%20437-557a1228%201228%200%2001345-137%2020%2020%200%200015-24c-3-10-13-17-24-15z'/%3e%3cpath%20d='M508%201252c67%20186%20176%20301%20299%20301%2090%200%20172-61%20237-169%206-9%203-22-7-27-9-6-22-3-27%207-58%2097-129%20149-203%20149-102%200-199-102-261-275a20%2020%200%2000-26-12%2020%2020%200%2000-12%2026zm610-36a1240%201240%200%200053-371c0-331-116-617-281-690-10-4-22%200-26%2010s0%2022%2010%2026c147%2065%20257%20336%20257%20654a1230%201230%200%2001-51%20359%2020%2020%200%200013%2025c10%203%2021-3%2025-13zm376-20a96%2096%200%2010-192%200%2096%2096%200%2000192%200zm-40%200a56%2056%200%2011-112%200%2056%2056%200%2001112%200zm-1241%2096a96%2096%200%20100-192%2096%2096%200%20000%20192zm0-40a56%2056%200%20110-112%2056%2056%200%20010%20112z'/%3e%3cpath%20d='M807%20252a96%2096%200%20100-192%2096%2096%200%20000%20192zm0-40a56%2056%200%20110-112%2056%2056%200%20010%20112zm14%20701a70%2070%200%2001-82-53%2070%2070%200%200153-82%2070%2070%200%200182%2053%2069%2069%200%2001-53%2082z'/%3e%3c/svg%3e";
const canaryIcon = "data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20viewBox='-10%20-10%20276%20276'%3e%3ccircle%20cx='128'%20cy='128'%20r='64'%20fill='%23fff'/%3e%3cpath%20fill='%23fbbc04'%20d='M96%20183.4A63.7%2063.7%200%200%201%2072.6%20160L17.2%2064A128%20128%200%200%200%20128%20256l55.4-96A64%2064%200%200%201%2096%20183.4Z'/%3e%3cpath%20fill='%23fdd663'%20d='M192%20128a63.7%2063.7%200%200%201-8.6%2032L128%20256A128%20128%200%200%200%20238.9%2064h-111a64%2064%200%200%201%2064%2064Z'/%3e%3ccircle%20cx='128'%20cy='128'%20r='52'%20fill='%23fbbc04'/%3e%3cpath%20fill='%23f29900'%20d='M96%2072.6a63.7%2063.7%200%200%201%2032-8.6h110.8a128%20128%200%200%200-221.7%200l55.5%2096A64%2064%200%200%201%2096%2072.6Z'/%3e%3c/svg%3e";
const chromeBetaIcon = "data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20fill='%23fff'%20viewBox='-190%20-190%205180%205180'%3e%3clinearGradient%20id='a'%20x1='2149'%20x2='4207'%20y1='4638'%20y2='1076'%20gradientUnits='userSpaceOnUse'%3e%3cstop%20offset='0'%20stop-color='%23fcc934'/%3e%3cstop%20offset='1'%20stop-color='%23fbbc04'/%3e%3c/linearGradient%3e%3clinearGradient%20id='b'%20x1='2823'%20x2='-214'%20y1='4031'%20y2='2278'%20gradientUnits='userSpaceOnUse'%3e%3cstop%20offset='0'%20stop-color='%231e8e3e'/%3e%3cstop%20offset='1'%20stop-color='%2334a853'/%3e%3c/linearGradient%3e%3clinearGradient%20id='c'%20x1='371'%20x2='4429'%20y1='1501'%20y2='1501'%20gradientUnits='userSpaceOnUse'%3e%3cstop%20offset='0'%20stop-color='%23d93025'/%3e%3cstop%20offset='1'%20stop-color='%23ea4335'/%3e%3c/linearGradient%3e%3cpath%20fill='%23202124'%20d='M4400%203676c-45%2080-93%20157-145%20228l541-228z'/%3e%3cpath%20d='M2399%203600a1200%201200%200%201%200-1%200z'/%3e%3cpath%20fill='url(%23a)'%20d='M3440%203001%202402%204800a2394%202394%200%200%200%202398-2400c0-421-112-835-323-1200H2400l-2%201a1198%201198%200%200%201%201041%201800z'/%3e%3cpath%20fill='url(%23b)'%20d='M1360%203001%20322%201202a2398%202398%200%200%200%202078%203598l1038-1799v-1a1199%201199%200%200%201-2078%202z'/%3e%3cpath%20fill='%231a73e8'%20d='M2399%203351a950%20950%200%201%200%200-1900%20950%20950%200%200%200%200%201900z'/%3e%3cpath%20fill='url(%23c)'%20d='M2399%201200h2076a2396%202396%200%200%200-4153%202l1038%201800h2a1200%201200%200%200%201%201037-1802z'/%3e%3cpath%20fill='%233c4043'%20d='M1575%202250c0-40%2033-74%2075-74h3149v1500H1649a75%2075%200%200%201-74-75z'/%3e%3cpath%20d='M2442%202883a199%20199%200%200%200%20115-185c1-44-10-87-34-124a244%20244%200%200%200-94-86c-38-20-81-30-125-30h-317v880h334c46%201%2090-10%20131-32a250%20250%200%200%200%20135-220%20216%20216%200%200%200-145-203zm-40%20325c-23%2014-50%2022-76%2021h-226v-286h222c28%200%2055%206%2079%2020a146%20146%200%200%201%2070%20123c0%2024-6%2047-17%2068-13%2022-30%2041-53%2054zm25-444a140%20140%200%200%201-124%2070h-203v-267h209c25%200%2050%207%2071%2020a140%20140%200%200%201%2065%20113c0%2022-6%2044-18%2063zm733-40a296%20296%200%200%200-160-42%20292%20292%200%200%200-264%20171%20389%20389%200%200%200%203%20340%20300%20300%200%200%200%20270%20164%20300%20300%200%200%200%20276-167l3-5-96-48-3%205c-40%2072-100%20110-176%20110a192%20192%200%200%201-168-95%20240%20240%200%200%201-32-112h481l1-4a377%20377%200%200%200-32-200%20277%20277%200%200%200-103-117zm-160%2064c35-1%2070%208%20100%2026%2025%2015%2045%2037%2059%2064%2010%2022%2018%2045%2020%2070h-360c10-44%2031-84%2063-114s74-47%20118-46zm657%20462a76%2076%200%200%201-62-25c-12-15-18-40-18-73v-340h150v-105h-150v-175h-114v175h-108v105h108v364c0%2027%203%2053%2012%2078a160%20160%200%200%200%2094%2090c23%2010%2047%2014%2071%2013%2017%200%2033%200%2050-3%2012-3%2025-7%2037-13l4-1v-115l-9%205c-20%2013-42%2020-65%2020zm656-500a283%20283%200%200%200-200-68c-50%200-100%2010-144%2033-40%2020-76%2051-101%2090l-3%204%2086%2066%203-5c17-26%2040-47%2067-61%2028-15%2060-22%2092-22%2043-2%2085%2013%20118%2040a142%20142%200%200%201%2046%20108v29a418%20418%200%200%200-170-36c-75%200-139%2019-189%2056s-75%2092-75%20160a205%20205%200%200%200%20114%20186%20268%20268%200%200%200%20245-5c30-17%2056-41%2076-70v83h108v-401c0-78-25-141-73-186zm-224%20504c-34%202-67-10-94-30a96%2096%200%200%201-36-79%20108%20108%200%200%201%2040-88c28-22%2070-34%20124-34a315%20315%200%200%201%20155%2042%20198%20198%200%200%201-95%20164c-28%2017-60%2026-94%2025z'/%3e%3c/svg%3e";
const chromeTestIcon = "" + new URL("browser-chrome-for-testing_x16-BtMPRNav.svg", import.meta.url).href;
const chromiumIcon = "data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20viewBox='-10%20-10%20276%20276'%3e%3ccircle%20cx='128'%20cy='128'%20r='64'%20fill='%23fff'/%3e%3cpath%20fill='%23669df6'%20d='M96%20183.4A63.7%2063.7%200%200%201%2072.6%20160L17.2%2064A128%20128%200%200%200%20128%20256l55.4-96A64%2064%200%200%201%2096%20183.4Z'/%3e%3cpath%20fill='%23aecbfa'%20d='M192%20128a63.7%2063.7%200%200%201-8.6%2032L128%20256A128%20128%200%200%200%20238.9%2064h-111a64%2064%200%200%201%2064%2064Z'/%3e%3ccircle%20cx='128'%20cy='128'%20r='52'%20fill='%231a73e8'/%3e%3cpath%20fill='%231967d2'%20d='M96%2072.6a63.7%2063.7%200%200%201%2032-8.6h110.8a128%20128%200%200%200-221.7%200l55.5%2096A64%2064%200%200%201%2096%2072.6Z'/%3e%3c/svg%3e";
const edgeBetaIcon = "" + new URL("edge-beta-hbvj7tXK.png", import.meta.url).href;
const edgeCanaryIcon = "" + new URL("edge-canary-DUj6lpZE.png", import.meta.url).href;
const edgeDevIcon = "" + new URL("edge-dev-B3yGmLPt.png", import.meta.url).href;
const firefoxNightlyIcon = "" + new URL("firefox-nightly-Cp5nfeDT.svg", import.meta.url).href;
const firefoxDeveloperEditionIcon = "" + new URL("firefox-developer-edition-CK_dVq6z.svg", import.meta.url).href;
const webKitIcon = "data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20xmlns:x='http://www.w3.org/1999/xlink'%20fill='%23fff'%20viewBox='0%200%205940%205940'%3e%3clinearGradient%20id='b'%20x1='50%25'%20x2='50%25'%20y1='0%25'%20y2='100%25'%3e%3cstop%20offset='0%25'%20stop-color='%2334aadc'/%3e%3cstop%20offset='100%25'%20stop-color='%23007aff'/%3e%3c/linearGradient%3e%3cpath%20id='a'%20d='M2354%204728L738%203458a611%20611%200%2001-253-480c0-180%2090-352%20253-480l1616-1272a1003%201003%200%2001616-199c233%200%20450%2072%20616%20200l1616%201270a611%20611%200%2001253%20480c0%20182-90%20353-253%20480L3586%204727a996%20996%200%2001-616%20201c-233%200-450-70-616-200z'/%3e%3cpath%20id='c'%20d='M2354%203928L738%202658a611%20611%200%2001-253-480c0-180%2090-352%20253-480L2354%20426a1003%201003%200%2001616-199c233%200%20450%2072%20616%20200l1616%201270a611%20611%200%2001253%20480c0%20182-90%20353-253%20480L3586%203927a996%20996%200%2001-616%20201c-233%200-450-70-616-200z'/%3e%3cpath%20fill='%23ff9d00'%20d='M5200%203280a590%20590%200%20010%20973L3590%205517a1060%201060%200%2001-1240%200L740%204255a590%20590%200%20011-974l1610-1264c342-266%20898-266%201240%200l1609%201263z'/%3e%3cuse%20fill='%23fc0'%20x:href='%23a'/%3e%3cuse%20fill='url(%23b)'%20x:href='%23c'/%3e%3cpath%20d='M2976%203395c825%200%201494-547%201494-1222%200-673-670-1223-1494-1223-826%200-1496%20550-1496%201223%200%20675%20670%201222%201495%201222zm0%20126c-910%200-1650-603-1650-1348%200-744%20740-1347%201650-1347s1648%20604%201648%201347c0%20745-738%201348-1648%201348z'/%3e%3cpath%20fill='%238cc8f6'%20d='M3086%202692c82-10%20163-34%20237-70l470%20194-246-370a430%20430%200%20000-546l246-370-443%20183-13%20195a330%20330%200%2001-12%20538%20514%20514%200%2001-160%2085l-80%20160zm-220-1036c-137%2018-240%2070-240%2070l-470-194%20246%20370a431%20431%200%20000%20545l-246%20370%20454-187%2010-183a330%20330%200%20013-546c45-40%20146-82%20160-85l80-160z'/%3e%3cpath%20fill-rule='evenodd'%20d='M2750%202133l-63%201110%20507-1030%2072-1107-516%201027zm-7%20897l406-824-356-65-50%20890z'/%3e%3c/svg%3e";
const genericBrowserLogo = "" + new URL("generic-browser-DnWpIeBK.svg", import.meta.url).href;
const allBrowsersIcons = {
  "Electron": electronIcon,
  "Chrome": chromeIcon,
  "Chrome Beta": chromeBetaIcon,
  "Chrome Canary": canaryIcon,
  "Chrome for Testing": chromeTestIcon,
  "Firefox": firefoxIcon,
  "Edge": edgeIcon,
  "Chromium": chromiumIcon,
  "Firefox Nightly": firefoxNightlyIcon,
  "Firefox Developer Edition": firefoxDeveloperEditionIcon,
  "Edge Canary": edgeCanaryIcon,
  "Edge Beta": edgeBetaIcon,
  "Edge Dev": edgeDevIcon,
  "WebKit": webKitIcon,
  "generic": genericBrowserLogo
};
const _hoisted_1 = ["data-browser-id", "onClick"];
const _hoisted_2 = ["src"];
const _hoisted_3 = { class: "grow" };
const _hoisted_4 = { class: "font-normal mr-[20px] text-gray-500 text-[14px] filter whitespace-nowrap group-focus-within:mix-blend-luminosity group-hover:mix-blend-luminosity" };
const _hoisted_5 = { key: 0 };
const _hoisted_6 = { class: "flex h-full items-center align-middle" };
const _hoisted_7 = {
  key: 0,
  "data-cy": "top-nav-browser-list-selected-item"
};
const _hoisted_8 = {
  key: 1,
  class: "h-[16px] relative"
};
const _hoisted_9 = { class: "text-center p-2 text-gray-300 text-[14px] leading-[20px]" };
const _hoisted_10 = /* @__PURE__ */ createBaseVNode("div", { class: "font-medium text-white mb-2" }, " Unsupported browser ", -1);
const _sfc_main$1 = /* @__PURE__ */ defineComponent({
  __name: "VerticalBrowserListItems",
  props: {
    selectable: { type: Boolean, default: false },
    gql: null,
    specPath: { default: "" }
  },
  setup(__props) {
    const props = __props;
    const { t } = useI18n();
    gql`
fragment VerticalBrowserListItems on CurrentProject {
  id
  browsers {
    id
    isSelected
    displayName
    version
    majorVersion
    isVersionSupported
    warning
    disabled
  }
}
`;
    gql`
mutation VerticalBrowserListItems_SetBrowser($id: ID!, $specPath: String) {
  launchpadSetBrowser(id: $id) {
    id
    ...VerticalBrowserListItems
  }
  launchOpenProject(specPath: $specPath) {
    id
  }
}
`;
    const browsers = computed(() => {
      return (props.gql.browsers ?? []).slice().sort((a, b) => a.displayName > b.displayName ? 1 : -1);
    });
    const setBrowser = useMutation(VerticalBrowserListItems_SetBrowserDocument);
    const handleBrowserChoice = async (browser) => {
      if (browser.disabled || !browser.isVersionSupported || browser.isSelected) {
        return;
      }
      const mutation = { id: browser.id, specPath: props.specPath };
      await setBrowser.executeMutation(mutation);
    };
    return (_ctx, _cache) => {
      const _component_i_cy_circle_check_x24 = __unplugin_components_0;
      const _component_i_cy_circle_bg_question_mark_x16 = __unplugin_components_0$1;
      return props.gql ? (openBlock(true), createElementBlock(Fragment, { key: 0 }, renderList(unref(browsers), (browser) => {
        return openBlock(), createElementBlock("li", {
          key: browser.id,
          class: normalizeClass(["border-b border-transparent cursor-pointer flex border-b-gray-50 border-[1px] min-w-[240px] py-[12px] px-[16px] transition-colors duration-300 group focus-within-default", {
            "bg-jade-50": browser.isSelected,
            "hover:bg-indigo-50 focus-within:bg-indigo-50": !browser.isSelected && !browser.disabled && browser.isVersionSupported,
            "bg-gray-50 cursor-not-allowed": browser.disabled || !browser.isVersionSupported,
            "cursor-pointer": !browser.disabled && browser.isVersionSupported
          }]),
          "data-cy": "top-nav-browser-list-item",
          "data-browser-id": browser.id,
          onClick: ($event) => handleBrowserChoice(browser)
        }, [
          createBaseVNode("img", {
            class: normalizeClass(["mr-[16px] min-w-[26px] w-[26px]", { "filter grayscale": browser.disabled || !browser.isVersionSupported }]),
            src: unref(allBrowsersIcons)[browser.displayName] || unref(allBrowsersIcons).generic,
            alt: ""
          }, null, 10, _hoisted_2),
          createBaseVNode("div", _hoisted_3, [
            createBaseVNode("div", null, [
              createBaseVNode("button", {
                class: normalizeClass(["font-medium box-border focus:outline-none", {
                  "text-indigo-500 group-hover:text-indigo-700": !browser.isSelected && !browser.disabled && browser.isVersionSupported,
                  "text-jade-700": browser.isSelected,
                  "text-gray-500": browser.disabled || !browser.isVersionSupported
                }])
              }, toDisplayString(browser.displayName), 3),
              createBaseVNode("div", _hoisted_4, [
                createTextVNode(toDisplayString(unref(t)("topNav.version")) + " " + toDisplayString(browser.majorVersion) + " ", 1),
                browser.version === "unsupported" ? (openBlock(), createElementBlock("span", _hoisted_5, " (" + toDisplayString(unref(t)("topNav.versionUnsupported")) + ") ", 1)) : createCommentVNode("", true)
              ])
            ])
          ]),
          createBaseVNode("div", null, [
            createBaseVNode("div", _hoisted_6, [
              browser.isSelected ? (openBlock(), createElementBlock("div", _hoisted_7, [
                createVNode(_component_i_cy_circle_check_x24, { class: "h-[24px] w-[24px] icon-dark-jade-100 icon-light-jade-500" })
              ])) : !browser.isVersionSupported ? (openBlock(), createElementBlock("div", _hoisted_8, [
                createVNode(_sfc_main$2, null, {
                  popper: withCtx(() => [
                    createBaseVNode("div", _hoisted_9, [
                      _hoisted_10,
                      createTextVNode(" " + toDisplayString(browser.warning), 1)
                    ])
                  ]),
                  default: withCtx(() => [
                    createVNode(_component_i_cy_circle_bg_question_mark_x16, {
                      class: "icon-dark-gray-700 icon-light-gray-200",
                      "data-cy": "unsupported-browser-tooltip-trigger"
                    })
                  ]),
                  _: 2
                }, 1024)
              ])) : createCommentVNode("", true)
            ])
          ])
        ], 10, _hoisted_1);
      }), 128)) : createCommentVNode("", true);
    };
  }
});
const __default__ = defineComponent({
  inheritAttrs: true
});
const _sfc_main = /* @__PURE__ */ defineComponent({
  ...__default__,
  __name: "HideDuringScreenshot",
  setup(__props) {
    const screenshotStore = useScreenshotStore();
    const classes = computed(() => {
      return {
        "hidden": screenshotStore.isScreenshotting
      };
    });
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", {
        class: normalizeClass([unref(classes), "p-0 m-0"])
      }, [
        renderSlot(_ctx.$slots, "default")
      ], 2);
    };
  }
});
const _imports_0 = "" + new URL("cypress-dark-BaptZgkl.png", import.meta.url).href;
export {
  BookIcon as B,
  _sfc_main$1 as _,
  allBrowsersIcons as a,
  _imports_0 as b,
  _sfc_main as c
};
