// @packages/app/dist/assets/PromoHeader.vue_vue_type_script_setup_true_lang-BXJKbPFz.js
import { o as openBlock, i as createElementBlock, a as createBaseVNode, d as defineComponent, g as useI18n, l as computed, F as Fragment, B as renderList, c as createBlock, n as normalizeClass, E as resolveDynamicComponent, t as toDisplayString, V as createTextVNode, b as unref, h as ref, f as createVNode, w as withCtx, m as renderSlot, bq as TransitionGroup, J as gql, aj as useMutation, aV as nanoid, aD as useQuery, bh as useSlots, j as createCommentVNode } from "./index-aHhBZXEW.js";
import { aB as Promo_PromoSeenDocument, aC as PromoDocument } from "./Tooltip.vue_vue_type_style_index_0_lang-CSb17MEq.js";
const _hoisted_1$8 = {
  style: { "min-width": "12px", "min-height": "12px" },
  width: "12",
  height: "12",
  viewBox: "0 0 12 12",
  fill: "none",
  xmlns: "http://www.w3.org/2000/svg"
};
const _hoisted_2$6 = /* @__PURE__ */ createBaseVNode("path", {
  d: "M3.17157 8.82843C4.73367 10.3905 7.26633 10.3905 8.82843 8.82843C10.3905 7.26633 10.3905 4.73367 8.82843 3.17157M3.17157 8.82843C1.60948 7.26633 1.60948 4.73367 3.17157 3.17157C4.73367 1.60948 7.26633 1.60948 8.82843 3.17157M3.17157 8.82843L8.82843 3.17157",
  stroke: "#BFC2D4",
  "stroke-width": "2",
  class: "icon-dark"
}, null, -1);
const _hoisted_3$6 = [
  _hoisted_2$6
];
function render$3(_ctx, _cache) {
  return openBlock(), createElementBlock("svg", _hoisted_1$8, _hoisted_3$6);
}
const SkippedIcon = { name: "cy-status-skipped_x12", render: render$3 };
const _hoisted_1$7 = {
  style: { "min-width": "12px", "min-height": "12px" },
  width: "12",
  height: "12",
  viewBox: "0 0 12 12",
  fill: "none",
  xmlns: "http://www.w3.org/2000/svg"
};
const _hoisted_2$5 = /* @__PURE__ */ createBaseVNode("path", {
  d: "M9 3L5 9L3 6.5",
  stroke: "#1FA971",
  "stroke-width": "2",
  "stroke-linecap": "round",
  "stroke-linejoin": "round",
  class: "icon-dark"
}, null, -1);
const _hoisted_3$5 = [
  _hoisted_2$5
];
function render$2(_ctx, _cache) {
  return openBlock(), createElementBlock("svg", _hoisted_1$7, _hoisted_3$5);
}
const PassedIcon = { name: "cy-status-passed_x12", render: render$2 };
const _hoisted_1$6 = {
  style: { "min-width": "12px", "min-height": "12px" },
  width: "12",
  height: "12",
  viewBox: "0 0 12 12",
  fill: "none",
  xmlns: "http://www.w3.org/2000/svg"
};
const _hoisted_2$4 = /* @__PURE__ */ createBaseVNode("path", {
  d: "M3.70711 2.29289C3.31658 1.90237 2.68342 1.90237 2.29289 2.29289C1.90237 2.68342 1.90237 3.31658 2.29289 3.70711L3.70711 2.29289ZM8.29289 9.70711C8.68342 10.0976 9.31658 10.0976 9.70711 9.70711C10.0976 9.31658 10.0976 8.68342 9.70711 8.29289L8.29289 9.70711ZM9.70711 3.70711C10.0976 3.31658 10.0976 2.68342 9.70711 2.29289C9.31658 1.90237 8.68342 1.90237 8.29289 2.29289L9.70711 3.70711ZM2.29289 8.29289C1.90237 8.68342 1.90237 9.31658 2.29289 9.70711C2.68342 10.0976 3.31658 10.0976 3.70711 9.70711L2.29289 8.29289ZM2.29289 3.70711L8.29289 9.70711L9.70711 8.29289L3.70711 2.29289L2.29289 3.70711ZM8.29289 2.29289L2.29289 8.29289L3.70711 9.70711L9.70711 3.70711L8.29289 2.29289Z",
  fill: "#E45770",
  class: "icon-dark"
}, null, -1);
const _hoisted_3$4 = [
  _hoisted_2$4
];
function render$1(_ctx, _cache) {
  return openBlock(), createElementBlock("svg", _hoisted_1$6, _hoisted_3$4);
}
const FailedIcon = { name: "cy-status-failed_x12", render: render$1 };
const _hoisted_1$5 = {
  style: { "min-width": "12px", "min-height": "12px" },
  width: "12",
  height: "12",
  viewBox: "0 0 12 12",
  fill: "none",
  xmlns: "http://www.w3.org/2000/svg"
};
const _hoisted_2$3 = /* @__PURE__ */ createBaseVNode("path", {
  d: "M4.47023 2.3053C3.5245 2.69703 2.7282 3.44956 2.30552 4.47001C1.46014 6.51092 2.42932 8.85072 4.47023 9.69609C6.51114 10.5415 8.85094 9.57229 9.69631 7.53138C10.119 6.51092 10.088 5.41575 9.69631 4.47001C9.30457 3.52428 8.55205 2.72798 7.5316 2.3053",
  class: "icon-dark",
  stroke: "#9095AD",
  "stroke-width": "2"
}, null, -1);
const _hoisted_3$3 = [
  _hoisted_2$3
];
function render(_ctx, _cache) {
  return openBlock(), createElementBlock("svg", _hoisted_1$5, _hoisted_3$3);
}
const PendingIcon = { name: "cy-status-pending_x12", render };
const _hoisted_1$4 = { class: "bg-white border rounded flex border-gray-200 h-6 text-gray-700 text-[14px] leading-[20px]" };
const _hoisted_2$2 = ["title", "data-cy"];
const _hoisted_3$2 = { class: "sr-only" };
const _sfc_main$4 = /* @__PURE__ */ defineComponent({
  __name: "ResultCounts",
  props: {
    totalPassed: null,
    totalFailed: null,
    totalPending: null,
    totalSkipped: null,
    order: null
  },
  setup(__props) {
    const props = __props;
    const { t } = useI18n();
    const ORDER_MAP = computed(() => {
      return {
        "SKIPPED": {
          value: props.totalSkipped,
          class: "icon-dark-gray-400",
          icon: SkippedIcon,
          name: t("runs.results.skipped")
        },
        "PENDING": {
          value: props.totalPending,
          class: "icon-dark-gray-400 icon-light-white",
          icon: PendingIcon,
          name: t("runs.results.pending")
        },
        "PASSED": {
          value: props.totalPassed,
          class: "icon-dark-jade-400",
          icon: PassedIcon,
          name: t("runs.results.passed")
        },
        "FAILED": {
          value: props.totalFailed,
          class: "icon-dark-red-400",
          icon: FailedIcon,
          name: t("runs.results.failed")
        }
      };
    });
    const results = computed(() => {
      const order = props.order || ["SKIPPED", "PENDING", "PASSED", "FAILED"];
      return order.map((status) => ORDER_MAP.value[status]);
    });
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", _hoisted_1$4, [
        (openBlock(true), createElementBlock(Fragment, null, renderList(unref(results), (result, i) => {
          return openBlock(), createElementBlock("div", {
            key: i,
            class: "flex px-2 items-center",
            title: result.name,
            "data-cy": `runResults-${result.name}-count`
          }, [
            (openBlock(), createBlock(resolveDynamicComponent(result.icon), {
              class: normalizeClass(["mt-px h-[12px] mr-1 w-[12px]", result.class]),
              "aria-hidden": "true"
            }, null, 8, ["class"])),
            createBaseVNode("span", _hoisted_3$2, toDisplayString(result.name), 1),
            createTextVNode(" " + toDisplayString(result.value), 1)
          ], 8, _hoisted_2$2);
        }), 128))
      ]);
    };
  }
});
const _hoisted_1$3 = { class: "relative" };
const _sfc_main$3 = /* @__PURE__ */ defineComponent({
  __name: "Slideshow",
  setup(__props) {
    const step = ref(0);
    const decrement = () => {
      step.value--;
    };
    const increment = () => {
      step.value++;
    };
    const reset = () => {
      step.value = 0;
    };
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", _hoisted_1$3, [
        createVNode(TransitionGroup, {
          "enter-from-class": "opacity-0",
          "enter-active-class": "transition duration-300 ease-out",
          "enter-to-class": "opacity-100",
          "leave-from-class": "opacity-100 absolute",
          "leave-active-class": "transition duration-300 ease-in absolute",
          "leave-to-class": "opacity-0 absolute"
        }, {
          default: withCtx(() => [
            renderSlot(_ctx.$slots, "default", {
              key: step.value,
              step: step.value,
              goBack: decrement,
              goForward: increment,
              reset
            })
          ]),
          _: 3
        })
      ]);
    };
  }
});
const _hoisted_1$2 = { class: "flex flex-col items-center border border-gray-100 rounded-lg shadow-[0_0_20px_rgba(0,0,0,0.08)]" };
const _sfc_main$2 = /* @__PURE__ */ defineComponent({
  __name: "Promo",
  props: {
    campaign: null,
    medium: null,
    cohort: null,
    instanceId: null
  },
  setup(__props) {
    const props = __props;
    gql`
fragment _Promo on Query {
  currentProject {
    id
    savedState
  }
}
`;
    gql`
query Promo {
  ..._Promo
}
`;
    gql`
mutation Promo_PromoSeen($campaign: String!, $messageId: String!, $medium: String!, $cohort: String) {
  recordEvent(campaign: $campaign, messageId: $messageId, medium: $medium, cohort: $cohort)
}
`;
    const promoSeenMutation = useMutation(Promo_PromoSeenDocument);
    const promoInstanceId = props.instanceId || nanoid();
    useQuery({ query: PromoDocument }).then((queryResult) => {
      var _a;
      const value = queryResult.data.value;
      if ((_a = value == null ? void 0 : value.currentProject) == null ? void 0 : _a.savedState) {
        promoSeenMutation.executeMutation({
          campaign: props.campaign,
          medium: props.medium,
          cohort: props.cohort ?? null,
          messageId: promoInstanceId
        });
      }
    });
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", _hoisted_1$2, [
        renderSlot(_ctx.$slots, "header"),
        createVNode(_sfc_main$3, null, {
          default: withCtx(({ step, goBack, goForward, reset }) => [
            renderSlot(_ctx.$slots, "cards", {
              step,
              goBack,
              goForward,
              reset
            })
          ]),
          _: 3
        })
      ]);
    };
  }
});
const _hoisted_1$1 = { class: "text-xl font-semibold text-gray-900" };
const _hoisted_2$1 = { class: "text-gray-700" };
const _hoisted_3$1 = {
  key: 0,
  class: "row-end-[span_2] xl:col-start-2"
};
const _hoisted_4$1 = {
  key: 1,
  class: "self-end"
};
const _sfc_main$1 = /* @__PURE__ */ defineComponent({
  __name: "PromoCard",
  props: {
    title: null,
    body: null
  },
  setup(__props) {
    const slots = useSlots();
    const gridClasses = computed(() => {
      const classes = ["grid", "grid-cols-[480px]", "gap-y-[16px]"];
      if (slots.image) {
        classes.push("xl:grid-cols-[300px_470px]", "gap-x-[100px]");
      } else {
        classes.push("xl:grid-cols-[744px]");
      }
      return classes;
    });
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", {
        class: normalizeClass(["p-[40px]", unref(gridClasses)])
      }, [
        createBaseVNode("div", null, [
          createBaseVNode("h2", _hoisted_1$1, toDisplayString(__props.title), 1),
          createBaseVNode("p", _hoisted_2$1, toDisplayString(__props.body), 1),
          renderSlot(_ctx.$slots, "content")
        ]),
        unref(slots).image ? (openBlock(), createElementBlock("div", _hoisted_3$1, [
          renderSlot(_ctx.$slots, "image")
        ])) : createCommentVNode("", true),
        unref(slots).action ? (openBlock(), createElementBlock("div", _hoisted_4$1, [
          renderSlot(_ctx.$slots, "action")
        ])) : createCommentVNode("", true)
      ], 2);
    };
  }
});
const _hoisted_1 = { class: "flex flex-col bg-gray-50 border-gray-100 py-[16px] px-[24px] rounded-t-[8px] gap-16px items-start w-full" };
const _hoisted_2 = { class: "flex flex-col xl:flex-row w-full items-start xl:items-center gap-[16px]" };
const _hoisted_3 = { class: "flex flex-col flex-grow justify-center" };
const _hoisted_4 = {
  "data-cy": "promo-header-title",
  class: "font-md font-medium"
};
const _hoisted_5 = {
  key: 0,
  "data-cy": "promo-header-description",
  class: "text-sm font-normal"
};
const _hoisted_6 = { key: 0 };
const _hoisted_7 = {
  key: 0,
  class: "pt-[24px] w-full"
};
const _sfc_main = /* @__PURE__ */ defineComponent({
  __name: "PromoHeader",
  props: {
    title: null
  },
  setup(__props) {
    const slots = useSlots();
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", _hoisted_1, [
        createBaseVNode("div", _hoisted_2, [
          createBaseVNode("div", _hoisted_3, [
            createBaseVNode("div", _hoisted_4, toDisplayString(__props.title), 1),
            unref(slots).description ? (openBlock(), createElementBlock("div", _hoisted_5, [
              renderSlot(_ctx.$slots, "description")
            ])) : createCommentVNode("", true)
          ]),
          unref(slots).control ? (openBlock(), createElementBlock("div", _hoisted_6, [
            renderSlot(_ctx.$slots, "control")
          ])) : createCommentVNode("", true)
        ]),
        unref(slots).content ? (openBlock(), createElementBlock("div", _hoisted_7, [
          renderSlot(_ctx.$slots, "content")
        ])) : createCommentVNode("", true)
      ]);
    };
  }
});
export {
  _sfc_main$4 as _,
  _sfc_main as a,
  _sfc_main$1 as b,
  _sfc_main$2 as c
};
