// @packages/data-context/src/DataActions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataActions = void 0;
const actions_1 = require("./actions");
const ErrorActions_1 = require("./actions/ErrorActions");
const EventCollectorActions_1 = require("./actions/EventCollectorActions");
const NotificationActions_1 = require("./actions/NotificationActions");
const VersionsActions_1 = require("./actions/VersionsActions");
class DataActions {
    constructor(ctx) {
        this.ctx = ctx;
        this._error = new ErrorActions_1.ErrorActions(this.ctx);
        this._file = new actions_1.FileActions(this.ctx);
        this._dev = new actions_1.DevActions(this.ctx);
        this._app = new actions_1.AppActions(this.ctx);
        this._auth = new actions_1.AuthActions(this.ctx);
        this._localSettings = new actions_1.LocalSettingsActions(this.ctx);
        this._wizard = new actions_1.WizardActions(this.ctx);
        this._project = new actions_1.ProjectActions(this.ctx);
        this._electron = new actions_1.ElectronActions(this.ctx);
        this._migration = new actions_1.MigrationActions(this.ctx);
        this._browser = new actions_1.BrowserActions(this.ctx);
        this._servers = new actions_1.ServersActions(this.ctx);
        this._versions = new VersionsActions_1.VersionsActions(this.ctx);
        this._eventCollector = new EventCollectorActions_1.EventCollectorActions(this.ctx);
        this._cohorts = new actions_1.CohortsActions(this.ctx);
        this._codegen = new actions_1.CodegenActions(this.ctx);
        this._notification = new NotificationActions_1.NotificationActions(this.ctx);
        this._cloudProject = new actions_1.CloudProjectActions(this.ctx);
    }
    get error() {
        return this._error;
    }
    get file() {
        return this._file;
    }
    get dev() {
        return this._dev;
    }
    get app() {
        return this._app;
    }
    get auth() {
        return this._auth;
    }
    get localSettings() {
        return this._localSettings;
    }
    get wizard() {
        return this._wizard;
    }
    get project() {
        return this._project;
    }
    get electron() {
        return this._electron;
    }
    get migration() {
        return this._migration;
    }
    get browser() {
        return this._browser;
    }
    get servers() {
        return this._servers;
    }
    get versions() {
        return this._versions;
    }
    get eventCollector() {
        return this._eventCollector;
    }
    get cohorts() {
        return this._cohorts;
    }
    get codegen() {
        return this._codegen;
    }
    get notification() {
        return this._notification;
    }
    get cloudProject() {
        return this._cloudProject;
    }
}
exports.DataActions = DataActions;
