// @packages/server/lib/cloud/studio/telemetry/constants/bundle-lifecycle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BUNDLE_LIFECYCLE_TELEMETRY_GROUPS = exports.BUNDLE_LIFECYCLE_TELEMETRY_GROUP_NAMES = exports.BUNDLE_LIFECYCLE_MEASURES = exports.BUNDLE_LIFECYCLE_MEASURE_NAMES = exports.BUNDLE_LIFECYCLE_MARK_NAMES = void 0;
exports.BUNDLE_LIFECYCLE_MARK_NAMES = {
    BUNDLE_LIFECYCLE_START: 'bundle_lifecycle_start',
    POST_STUDIO_SESSION_START: 'post_studio_session_start',
    POST_STUDIO_SESSION_END: 'post_studio_session_end',
    ENSURE_STUDIO_BUNDLE_START: 'ensure_studio_bundle_start',
    ENSURE_STUDIO_BUNDLE_END: 'ensure_studio_bundle_end',
    STUDIO_MANAGER_SETUP_START: 'studio_manager_setup_start',
    STUDIO_MANAGER_SETUP_END: 'studio_manager_setup_end',
    STUDIO_PROTOCOL_GET_START: 'studio_protocol_get_start',
    STUDIO_PROTOCOL_GET_END: 'studio_protocol_get_end',
    STUDIO_PROTOCOL_PREPARE_START: 'studio_protocol_prepare_start',
    STUDIO_PROTOCOL_PREPARE_END: 'studio_protocol_prepare_end',
    BUNDLE_LIFECYCLE_END: 'bundle_lifecycle_end',
};
exports.BUNDLE_LIFECYCLE_MEASURE_NAMES = {
    BUNDLE_LIFECYCLE_DURATION: 'bundle_lifecycle_duration',
    POST_STUDIO_SESSION_DURATION: 'post_studio_session_duration',
    ENSURE_STUDIO_BUNDLE_DURATION: 'ensure_studio_bundle_duration',
    STUDIO_MANAGER_SETUP_DURATION: 'studio_manager_setup_duration',
    STUDIO_PROTOCOL_GET_DURATION: 'studio_protocol_get_duration',
    STUDIO_PROTOCOL_PREPARE_DURATION: 'studio_protocol_prepare_duration',
};
exports.BUNDLE_LIFECYCLE_MEASURES = {
    [exports.BUNDLE_LIFECYCLE_MEASURE_NAMES.BUNDLE_LIFECYCLE_DURATION]: [
        exports.BUNDLE_LIFECYCLE_MARK_NAMES.BUNDLE_LIFECYCLE_START,
        exports.BUNDLE_LIFECYCLE_MARK_NAMES.BUNDLE_LIFECYCLE_END,
    ],
    [exports.BUNDLE_LIFECYCLE_MEASURE_NAMES.POST_STUDIO_SESSION_DURATION]: [
        exports.BUNDLE_LIFECYCLE_MARK_NAMES.POST_STUDIO_SESSION_START,
        exports.BUNDLE_LIFECYCLE_MARK_NAMES.POST_STUDIO_SESSION_END,
    ],
    [exports.BUNDLE_LIFECYCLE_MEASURE_NAMES.ENSURE_STUDIO_BUNDLE_DURATION]: [
        exports.BUNDLE_LIFECYCLE_MARK_NAMES.ENSURE_STUDIO_BUNDLE_START,
        exports.BUNDLE_LIFECYCLE_MARK_NAMES.ENSURE_STUDIO_BUNDLE_END,
    ],
    [exports.BUNDLE_LIFECYCLE_MEASURE_NAMES.STUDIO_MANAGER_SETUP_DURATION]: [
        exports.BUNDLE_LIFECYCLE_MARK_NAMES.STUDIO_MANAGER_SETUP_START,
        exports.BUNDLE_LIFECYCLE_MARK_NAMES.STUDIO_MANAGER_SETUP_END,
    ],
    [exports.BUNDLE_LIFECYCLE_MEASURE_NAMES.STUDIO_PROTOCOL_GET_DURATION]: [
        exports.BUNDLE_LIFECYCLE_MARK_NAMES.STUDIO_PROTOCOL_GET_START,
        exports.BUNDLE_LIFECYCLE_MARK_NAMES.STUDIO_PROTOCOL_GET_END,
    ],
    [exports.BUNDLE_LIFECYCLE_MEASURE_NAMES.STUDIO_PROTOCOL_PREPARE_DURATION]: [
        exports.BUNDLE_LIFECYCLE_MARK_NAMES.STUDIO_PROTOCOL_PREPARE_START,
        exports.BUNDLE_LIFECYCLE_MARK_NAMES.STUDIO_PROTOCOL_PREPARE_END,
    ],
};
exports.BUNDLE_LIFECYCLE_TELEMETRY_GROUP_NAMES = {
    COMPLETE_BUNDLE_LIFECYCLE: 'complete-bundle-lifecycle',
};
exports.BUNDLE_LIFECYCLE_TELEMETRY_GROUPS = {
    [exports.BUNDLE_LIFECYCLE_TELEMETRY_GROUP_NAMES.COMPLETE_BUNDLE_LIFECYCLE]: [
        exports.BUNDLE_LIFECYCLE_MEASURE_NAMES.BUNDLE_LIFECYCLE_DURATION,
        exports.BUNDLE_LIFECYCLE_MEASURE_NAMES.POST_STUDIO_SESSION_DURATION,
        exports.BUNDLE_LIFECYCLE_MEASURE_NAMES.ENSURE_STUDIO_BUNDLE_DURATION,
        exports.BUNDLE_LIFECYCLE_MEASURE_NAMES.STUDIO_MANAGER_SETUP_DURATION,
        exports.BUNDLE_LIFECYCLE_MEASURE_NAMES.STUDIO_PROTOCOL_GET_DURATION,
        exports.BUNDLE_LIFECYCLE_MEASURE_NAMES.STUDIO_PROTOCOL_PREPARE_DURATION,
    ],
};
