// @packages/server/lib/cloud/studio/telemetry/constants/initialization.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.INITIALIZATION_TELEMETRY_GROUPS = exports.INITIALIZATION_TELEMETRY_GROUP_NAMES = exports.INITIALIZATION_MEASURES = exports.INITIALIZATION_MEASURE_NAMES = exports.INITIALIZATION_MARK_NAMES = void 0;
exports.INITIALIZATION_MARK_NAMES = Object.freeze({
    INITIALIZATION_START: 'initialization-start',
    INITIALIZATION_END: 'initialization-end',
    CAN_ACCESS_STUDIO_AI_START: 'can-access-studio-ai-start',
    CAN_ACCESS_STUDIO_AI_END: 'can-access-studio-ai-end',
    CONNECT_PROTOCOL_TO_BROWSER_START: 'connect-protocol-to-browser-start',
    CONNECT_PROTOCOL_TO_BROWSER_END: 'connect-protocol-to-browser-end',
    INITIALIZE_STUDIO_AI_START: 'initialize-studio-ai-start',
    INITIALIZE_STUDIO_AI_END: 'initialize-studio-ai-end',
});
exports.INITIALIZATION_MEASURE_NAMES = Object.freeze({
    INITIALIZATION_DURATION: 'initialization-duration',
    CAN_ACCESS_STUDIO_AI_DURATION: 'can-access-studio-ai-duration',
    CONNECT_PROTOCOL_TO_BROWSER_DURATION: 'connect-protocol-to-browser-duration',
    INITIALIZE_STUDIO_AI_DURATION: 'initialize-studio-ai-duration',
});
exports.INITIALIZATION_MEASURES = Object.freeze({
    [exports.INITIALIZATION_MEASURE_NAMES.INITIALIZATION_DURATION]: [
        exports.INITIALIZATION_MARK_NAMES.INITIALIZATION_START,
        exports.INITIALIZATION_MARK_NAMES.INITIALIZATION_END,
    ],
    [exports.INITIALIZATION_MEASURE_NAMES.CAN_ACCESS_STUDIO_AI_DURATION]: [
        exports.INITIALIZATION_MARK_NAMES.CAN_ACCESS_STUDIO_AI_START,
        exports.INITIALIZATION_MARK_NAMES.CAN_ACCESS_STUDIO_AI_END,
    ],
    [exports.INITIALIZATION_MEASURE_NAMES.CONNECT_PROTOCOL_TO_BROWSER_DURATION]: [
        exports.INITIALIZATION_MARK_NAMES.CONNECT_PROTOCOL_TO_BROWSER_START,
        exports.INITIALIZATION_MARK_NAMES.CONNECT_PROTOCOL_TO_BROWSER_END,
    ],
    [exports.INITIALIZATION_MEASURE_NAMES.INITIALIZE_STUDIO_AI_DURATION]: [
        exports.INITIALIZATION_MARK_NAMES.INITIALIZE_STUDIO_AI_START,
        exports.INITIALIZATION_MARK_NAMES.INITIALIZE_STUDIO_AI_END,
    ],
});
exports.INITIALIZATION_TELEMETRY_GROUP_NAMES = Object.freeze({
    INITIALIZE_STUDIO: 'initialize-studio',
});
exports.INITIALIZATION_TELEMETRY_GROUPS = Object.freeze({
    [exports.INITIALIZATION_TELEMETRY_GROUP_NAMES.INITIALIZE_STUDIO]: [
        exports.INITIALIZATION_MEASURE_NAMES.INITIALIZATION_DURATION,
        exports.INITIALIZATION_MEASURE_NAMES.CAN_ACCESS_STUDIO_AI_DURATION,
        exports.INITIALIZATION_MEASURE_NAMES.CONNECT_PROTOCOL_TO_BROWSER_DURATION,
        exports.INITIALIZATION_MEASURE_NAMES.INITIALIZE_STUDIO_AI_DURATION,
    ],
});
