// @packages/types/src/preferences.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.allowedKeys = exports.defaultPreferences = exports.NotifyCompletionStatuses = void 0;
exports.NotifyCompletionStatuses = ['passed', 'failed', 'cancelled', 'errored'];
exports.defaultPreferences = {
    autoScrollingEnabled: true,
    isSpecsListOpen: false,
    isSideNavigationOpen: true,
    desktopNotificationsEnabled: null,
    notifyWhenRunStarts: false,
    notifyWhenRunStartsFailing: true,
    notifyWhenRunCompletes: ['failed'],
};
exports.allowedKeys = [
    'appWidth',
    'appHeight',
    'appX',
    'appY',
    'autoScrollingEnabled',
    'banners',
    'browserWidth',
    'browserHeight',
    'browserX',
    'browserY',
    'isAppDevToolsOpen',
    'isBrowserDevToolsOpen',
    'reporterWidth',
    'specListWidth',
    'showedNewProjectBanner',
    'firstOpenedCypress',
    'showedStudioModal',
    'preferredOpener',
    'isSpecsListOpen',
    'firstOpened',
    'lastOpened',
    'lastProjectId',
    'lastTestCountsEvent',
    'promptsShown',
    'specFilter',
    'preferredEditorBinary',
    'desktopNotificationsEnabled',
    'dismissNotificationBannerUntil',
    'isSideNavigationOpen',
    'lastBrowser',
    'majorVersionWelcomeDismissed',
    'debugSlideshowComplete',
    'notifyWhenRunStarts',
    'notifyWhenRunStartsFailing',
    'notifyWhenRunCompletes',
];
