// @packages/proxy/lib/network-proxy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkProxy = void 0;
const telemetry_1 = require("../../telemetry");
const http_1 = require("./http");
class NetworkProxy {
    constructor(opts) {
        this.http = new http_1.Http(opts);
    }
    async addPendingBrowserPreRequest(preRequest) {
        await this.http.addPendingBrowserPreRequest(preRequest);
    }
    removePendingBrowserPreRequest(requestId) {
        this.http.removePendingBrowserPreRequest(requestId);
    }
    getPendingBrowserPreRequests() {
        return this.http.getPendingBrowserPreRequests();
    }
    addPendingUrlWithoutPreRequest(url) {
        this.http.addPendingUrlWithoutPreRequest(url);
    }
    updateServiceWorkerRegistrations(data) {
        this.http.updateServiceWorkerRegistrations(data);
    }
    updateServiceWorkerVersions(data) {
        this.http.updateServiceWorkerVersions(data);
    }
    updateServiceWorkerClientSideRegistrations(data) {
        this.http.updateServiceWorkerClientSideRegistrations(data);
    }
    handleServiceWorkerClientEvent(event) {
        this.http.handleServiceWorkerClientEvent(event);
    }
    async handleHttpRequest(req, res) {
        const span = telemetry_1.telemetry.startSpan({
            name: 'network:proxy:handleHttpRequest',
            opts: {
                attributes: {
                    'network:proxy:url': req.proxiedUrl,
                    'network:proxy:contentType': req.get('content-type'),
                },
            },
            isVerbose: true,
        });
        await this.http.handleHttpRequest(req, res, span).finally(() => {
            span === null || span === void 0 ? void 0 : span.end();
        });
    }
    async handleSourceMapRequest(req, res) {
        await this.http.handleSourceMapRequest(req, res);
    }
    setHttpBuffer(buffer) {
        this.http.setBuffer(buffer);
    }
    reset(options = { resetBetweenSpecs: false }) {
        this.http.reset(options);
    }
    setProtocolManager(protocolManager) {
        this.http.setProtocolManager(protocolManager);
    }
    setPreRequestTimeout(timeout) {
        this.http.setPreRequestTimeout(timeout);
    }
}
exports.NetworkProxy = NetworkProxy;
