// @packages/server/lib/cloud/network/http_error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpError = exports.HttpErrorKind = void 0;
const scrub_url_1 = require("../api/scrub_url");
exports.HttpErrorKind = 'HttpError';
class HttpError extends Error {
    constructor(message, url, status, statusText, responseBody, originalResponse) {
        super(message);
        this.url = url;
        this.status = status;
        this.statusText = statusText;
        this.responseBody = responseBody;
        this.originalResponse = originalResponse;
        this.kind = exports.HttpErrorKind;
    }
    static isHttpError(error) {
        return (error === null || error === void 0 ? void 0 : error.kind) === exports.HttpErrorKind && Boolean(error.originalResponse);
    }
    static async fromResponse(response) {
        const status = response.status;
        const statusText = response.statusText;
        const responseBody = await response.text();
        const scrubbedUrl = (0, scrub_url_1.scrubUrl)(response.url);
        return new HttpError(`${(0, scrub_url_1.scrubUrl)(response.url)} responded with ${status} ${statusText}`, scrubbedUrl, status, statusText, responseBody, response);
    }
}
exports.HttpError = HttpError;
